server/src/main/java/org/apache/druid/metadata/SqlSegmentsMetadataQuery.java

```diff
@@ -297,6 +315,66 @@ private CloseableIterator<DataSegment> retrieveSegments(
     );
   }
 
+  private CloseableIterator<DataSegment> retrieveSegments(
+      final String dataSource,
+      final List<String> segmentIds,
+      final boolean used
+  )
+  {
+    if (CollectionUtils.isNullOrEmpty(segmentIds)) {
+      return CloseableIterators.withEmptyBaggage(ImmutableList.<DataSegment>of().iterator());
+    }
+    int batchSize = 100;
```

Comment:
maybe this batchSize should be pushed to higher level, and configurable, Maybe the retreiveUnusedSegmentsIdAction